repair/repair.cc

```diff
@@ -1989,26 +2024,81 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
+                std::vector<sstring> dcs;
+                dcs.reserve(live_nodes_per_dc.size());
+                std::ranges::copy_if(topology.get_datacenters(), std::back_inserter(dcs), [&] (const auto& dc) {
+                    return dc != *source_dc && !live_nodes_per_dc[dc].empty();
+                });
+                if (!dcs.empty()) {
+                    std::uniform_int_distribution<int> dist(0, dcs.size() - 1);
+                    return dcs[dist(_random_engine)];
+                }
+                return sstring();
+            };
+            // Allow repairing in the source_dc only if there are enough replicas remaining
+            if (source_dc) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    const auto& nt_strat = dynamic_cast<const locator::network_topology_strategy&>(strat);
+                    size_t rf = nt_strat.get_replication_factor(*source_dc);
+                    auto lost = lost_nodes_per_dc[*source_dc];
+                    source_dc_for_keyspace = *source_dc;
+
+                    // Look for an alternative source_dc if it's unsafe to repair only
+                    // from the source_dc.
+                    if (rf == 0 || lost > 1 || (lost == 1 && rf == 1)) {
+                        alt_dc_for_keyspace = find_alternative_datacenter();
+                        auto msg = format("{}: it is unsafe to use source_dc={} to repair/rebuild keyspace={} since it lost {} nodes, rf={}", op, source_dc, keyspace_name, lost, rf);
+                        if (alt_dc_for_keyspace.empty()) {
+                            msg = fmt::format("{}: found no alternative datacenter", msg);
+                            if (source_dc.force()) {
+                                rlogger.warn("{}: using source_dc anyway according to the force option", msg);
+                            } else if (source_dc.user_provided()) {
+                                throw std::runtime_error(format("{}: Omit the source_dc option to use all nodes in the cluster, or use the --force option to enforce it", msg));
+                            } else {
+                                rlogger.warn("{}: falling back to repair using all nodes", msg);
+                                source_dc_for_keyspace = "";
+                            }
+                        }
+                    }
+                    break;
+                }
+                case locator::replication_strategy_type::everywhere_topology:
+                    // source_dc_live_nodes is not empty, so we can use any remaining nodes
+                    source_dc_for_keyspace = *source_dc;
+                    break;
+                case locator::replication_strategy_type::simple:
+                    // With simple strategy, we have no assurance that source_dc will contain
+                    // another replica for all token ranges.
```

Comment:
It is not obvious that empty source_dc_for_keyspace means allowing all replicas. I know source_dc_for_keyspace  is empty by default. How about setting source_dc_for_keyspace = "" explicitly and add a comment here.

Reply:
sure, no problem.

Reply:
done in v8